engines/director/score.cpp

```diff
@@ -1344,123 +1357,273 @@ void Score::playQueuedSound() {
 		// Unknown, some bytes - constant (refer to contuinity).
 	}
 
-	uint16 channelSize;
-	uint16 channelOffset;
+	// TODO How to work with channelData that gets overridden
+	// partically by channels, hence we keep it and read the score from left to right
+	// TODO Merge it with shared cast
+	memset(_channelData, 0, kChannelDataSize);
 
-	Frame *initial = new Frame(this, _numChannelsDisplayed);
-	// Push a frame at frame#0 position.
 	// This makes all indexing simpler
-	_frames.push_back(initial);
+	_frameOffsets.push_back(0);
+
+	_currentTempo = 0;
+	_currentPaletteId = CastMemberID(0, 0);
+
+	// Prepare frameOffsets
+	_version = version;
+	_frameOffsets.push_back(_framesStream->pos());
+
+	// In case number of frames not known, precompute them!
+	if (_numFrames == 0) {
+		debugC(1, kDebugLoading, "Score::loadFrames(): Precomputing total number of frames!");
+		// Calculate number of frames beforehand.
+		int frameCount = 1;
+		while (loadFrame(frameCount)) {
+			debugC(1, kDebugLoading, "Score::loadFrames(): Skipped over frame %d!", frameCount);
+			frameCount++;
+		}
+		_numFrames = frameCount;
+		debugC(1, kDebugLoading, "Score::loadFrames(): Calculated, total number of frames %d!", _numFrames);
 
-	// This is a representation of the channelData. It gets overridden
-	// partically by channels, hence we keep it and read the score from left to right
-	//
-	// TODO Merge it with shared cast
-	byte channelData[kChannelDataSize];
-	memset(channelData, 0, kChannelDataSize);
+		memset(_channelData, 0, kChannelDataSize); // Reset channel data
+	}
 
-	uint8 currentTempo = 0;
-	CastMemberID currentPaletteId = CastMemberID(0, 0);
+	loadFrame(1);
 
-	while (size != 0 && !stream.eos()) {
-		uint16 frameSize = stream.readUint16();
-		debugC(3, kDebugLoading, "++++++++++ score frame %d (frameSize %d) size %d", _frames.size(), frameSize, size);
-		if (debugChannelSet(8, kDebugLoading)) {
-			stream.hexdump(frameSize);
+	// Read over frame offset array and print each item
+	for (uint i = 0; i < _frameOffsets.size(); i++) {
+		debugC(1, kDebugLoading, "Score::loadFrames(): Frame %d, offset %ld", i, _frameOffsets[i]);
+	}
+
+	debugC(1, kDebugLoading, "Score::loadFrames(): Number of frames: %d, framesStreamSize: %d", _numFrames, _framesStreamSize);
+}
+
+bool Score::loadFrame(int frameNum) {
+	debugC(7, kDebugLoading, "****** A call to loadFrame %d", frameNum);
+
+	// Read existing frame (ie already visited)
+	if (frameNum < (int)_frameOffsets.size() - 1) {
+		// TODO: How _channelData be modified.
+		// As of now channelData is left unchanged, only rebuilt when there is jumps
+		// We have this frame already, seek and load
+		_framesStream->seek(_frameOffsets[frameNum]);
+		debugC(7, kDebugLoading, "****** Frame request %d, offset %ld", frameNum, _framesStream->pos());
+
+		if (readOneFrame()) {
+			loadCurrentFrameAction(); // Load actions of current frame!
+			setSpriteCasts(); // Set sprite casts for each sprite in frame!
```

Comment:
A lot of code duplication. Look, there is no difference in frame processing besides seeking.

So, write something like

```C++
if (frameNum < (int)_frameOffsets.size() - 1) {
  _framesStream->seek(_frameOffsets[frameNum]);
}

while (_frameOffsets.size() <= (uint)frameNum) {
  ....
 ```

because your first seeking will work exactly the same as you're coming from a previous frame.

However, as we discussed, you need to investigate what is happening with partial channel updates in this case. I suspect, that nothing, we just seek to this frame and read it. But testing is REQUIRED, we cannot do this blindly.